File: ios/FaceAuthApp/DetectFacesPlugin.swift
import Foundation
import VisionCamera
import TensorFlowLite
import CoreImage
import UIKit
import Accelerate

/**
 * VisionCamera v3 Frame Processor Plugin — iOS (Swift + Metal GPU delegate)
 *
 * Registered as "detectFaces" via the Objective-C bridge file.
 * Called from JS worklet: global.detectFaces(frame)
 *
 * Input:  CVPixelBuffer (kCVPixelFormatType_420YpCbCr8BiPlanarFullRange)
 * Output: { detections: [{ x, y, width, height, score, keypoints: [{x,y}×6] }] }
 */
@objc(DetectFacesPlugin)
public class DetectFacesPlugin: FrameProcessorPlugin {

    // MARK: – Constants
    private static let inputSize   = 128
    private static let numAnchors  = 896
    private static let numCoords   = 16
    private static let scoreThresh = Float(0.75)
    private static let iouThresh   = Float(0.3)
    private static let modelName   = "blazeface"

    // MARK: – Lazy singletons (first frame init)
    private lazy var interpreter: Interpreter? = {
        guard let modelPath = Bundle.main.path(forResource: Self.modelName, ofType: "tflite", inDirectory: "models")
        else { return nil }

        var options = Interpreter.Options()
        options.threadCount = 2

        // Use Metal GPU delegate on iOS for ~2× speedup
        let metalDelegate = MetalDelegate()
        do {
            return try Interpreter(modelPath: modelPath, delegates: [metalDelegate])
        } catch {
            // Fallback to CPU
            return try? Interpreter(modelPath: modelPath)
        }
    }()

    private lazy var anchors: [[Float]] = generateAnchors()

    // MARK: – Plugin entry point

    public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
        guard let interpreter = interpreter else { return ["detections": []] }

        let pixelBuffer = CMSampleBufferGetImageBuffer(frame.buffer)!

        guard let inputData = preprocess(pixelBuffer: pixelBuffer) else {
            return ["detections": []]
        }

        do {
            try interpreter.allocateTensors()
            try interpreter.copy(inputData, toInputAt: 0)
            try interpreter.invoke()

            let regressorsTensor = try interpreter.output(at: 0)
            let scoresTensor     = try interpreter.output(at: 1)

            let regressors = regressorsTensor.data.toArray(type: Float.self)
            let scores     = scoresTensor.data.toArray(type: Float.self)

            let detections = decode(
                regressors : regressors,
                scores     : scores,
                frameW     : frame.width,
                frameH     : frame.height
            )
            return ["detections": detections]

        } catch {
            return ["detections": []]
        }
    }

    // MARK: – Preprocessing

    private func preprocess(pixelBuffer: CVPixelBuffer) -> Data? {
        let ciImage = CIImage(cvPixelBuffer: pixelBuffer)
        let context = CIContext()

        let size = CGSize(width: Self.inputSize, height: Self.inputSize)
        let scaleX = size.width  / CGFloat(CVPixelBufferGetWidth(pixelBuffer))
        let scaleY = size.height / CGFloat(CVPixelBufferGetHeight(pixelBuffer))
        let scaled = ciImage.transformed(by: CGAffineTransform(scaleX: scaleX, y: scaleY))

        guard let cgImage = context.createCGImage(scaled, from: CGRect(origin: .zero, size: size))
        else { return nil }

        let channelCount = 3
        var rawBytes = [UInt8](repeating: 0, count: Self.inputSize * Self.inputSize * 4)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        guard let ctx = CGContext(
            data        : &rawBytes,
            width       : Self.inputSize,
            height      : Self.inputSize,
            bitsPerComponent: 8,
            bytesPerRow : Self.inputSize * 4,
            space       : colorSpace,
            bitmapInfo  : CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return nil }
        ctx.draw(cgImage, in: CGRect(origin: .zero, size: size))

        // RGBA → normalised RGB float32 [-1, 1]
        var floats = [Float32](repeating: 0, count: Self.inputSize * Self.inputSize * channelCount)
        for i in 0 ..< Self.inputSize * Self.inputSize {
            floats[i * 3]     = Float(rawBytes[i * 4])     / 127.5 - 1
            floats[i * 3 + 1] = Float(rawBytes[i * 4 + 1]) / 127.5 - 1
            floats[i * 3 + 2] = Float(rawBytes[i * 4 + 2]) / 127.5 - 1
        }
        return Data(bytes: floats, count: floats.count * MemoryLayout<Float32>.stride)
    }

    // MARK: – Decoding

    private func decode(regressors: [Float], scores: [Float], frameW: Int, frameH: Int) -> [[String: Any]] {
        let scaleX = Float(frameW) / Float(Self.inputSize)
        let scaleY = Float(frameH) / Float(Self.inputSize)
        var raw = [[String: Any]]()

        for i in 0 ..< Self.numAnchors {
            let score = sigmoid(scores[i])
            guard score >= Self.scoreThresh else { continue }

            let base = i * Self.numCoords
            let ax   = anchors[i][0]
            let ay   = anchors[i][1]
            let cx   = regressors[base]     / Float(Self.inputSize) + ax
            let cy   = regressors[base + 1] / Float(Self.inputSize) + ay
            let w    = regressors[base + 2] / Float(Self.inputSize)
            let h    = regressors[base + 3] / Float(Self.inputSize)

            var kps = [[String: Double]]()
            for k in 0 ..< 6 {
                kps.append([
                    "x": Double((regressors[base + 4 + k * 2]     / Float(Self.inputSize) + ax) * scaleX),
                    "y": Double((regressors[base + 4 + k * 2 + 1] / Float(Self.inputSize) + ay) * scaleY),
                ])
            }

            raw.append([
                "x":         Double((cx - w / 2) * scaleX),
                "y":         Double((cy - h / 2) * scaleY),
                "width":     Double(w * scaleX),
                "height":    Double(h * scaleY),
                "score":     Double(score),
                "keypoints": kps,
            ])
        }

        return nonMaxSuppression(raw)
    }

    private func nonMaxSuppression(_ dets: [[String: Any]]) -> [[String: Any]] {
        let sorted     = dets.sorted { ($0["score"] as! Double) > ($1["score"] as! Double) }
        var kept       = [[String: Any]]()
        var suppressed = [Bool](repeating: false, count: sorted.count)

        for i in sorted.indices {
            guard !suppressed[i] else { continue }
            kept.append(sorted[i])
            for j in (i + 1) ..< sorted.count {
                if !suppressed[j] && iou(sorted[i], sorted[j]) > Self.iouThresh {
                    suppressed[j] = true
                }
            }
        }
        return kept
    }

    private func iou(_ a: [String: Any], _ b: [String: Any]) -> Float {
        let ax1 = Float(a["x"] as! Double), ay1 = Float(a["y"] as! Double)
        let ax2 = ax1 + Float(a["width"] as! Double), ay2 = ay1 + Float(a["height"] as! Double)
        let bx1 = Float(b["x"] as! Double), by1 = Float(b["y"] as! Double)
        let bx2 = bx1 + Float(b["width"] as! Double), by2 = by1 + Float(b["height"] as! Double)

        let ix = max(0, min(ax2, bx2) - max(ax1, bx1))
        let iy = max(0, min(ay2, by2) - max(ay1, by1))
        let inter = ix * iy
        let union = (ax2-ax1)*(ay2-ay1) + (bx2-bx1)*(by2-by1) - inter
        return inter / (union + 1e-8)
    }

    private func sigmoid(_ x: Float) -> Float { 1 / (1 + exp(-x)) }

    // MARK: – Anchor generation (mirrors generate_anchors.py)

    private func generateAnchors() -> [[Float]] {
        let strides = [8, 16, 16, 16]
        var anchors = [[Float]]()
        let offset  = Float(0.5)

        for stride in strides {
            let fmW = (Self.inputSize + stride - 1) / stride
            let fmH = fmW
            for y in 0 ..< fmH {
                for x in 0 ..< fmW {
                    for _ in 0 ..< 2 {
                        anchors.append([
                            (Float(x) + offset) / Float(fmW),
                            (Float(y) + offset) / Float(fmH),
                            1, 1,
                        ])
                    }
                }
            }
        }
        return anchors
    }
}

// MARK: – Data extension helper
private extension Data {
    func toArray<T>(type: T.Type) -> [T] {
        withUnsafeBytes { Array($0.bindMemory(to: T.self)) }
    }
}

File: android/app/src/main/java/com/faceauthapp/frameprocessor/DetectFacesPlugin.kt
package com.faceauthapp.frameprocessor

import android.content.res.AssetFileDescriptor
import android.graphics.Bitmap
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import org.tensorflow.lite.Interpreter
import java.io.ByteArrayOutputStream
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * VisionCamera v3 Frame Processor Plugin — runs BlazeFace on every camera frame.
 *
 * Registered as "detectFaces" in FaceDetectionPackage.
 * Called from JS worklet: global.detectFaces(frame)
 *
 * Returns: { detections: [{ x, y, width, height, score, keypoints: [{x,y}×6] }] }
 */
class DetectFacesPlugin(proxy: VisionCameraProxy, options: Map<String, Any>?) :
    FrameProcessorPlugin(proxy, options) {

    companion object {
        private const val INPUT_SIZE    = 128
        private const val NUM_ANCHORS   = 896
        private const val NUM_COORDS    = 16  // 4 box + 6 keypoints × 2
        private const val SCORE_THRESH  = 0.75f
        private const val IOU_THRESH    = 0.3f
        private const val MODEL_FILE    = "models/blazeface.tflite"
    }

    private val interpreter: Interpreter by lazy {
        Interpreter(loadModelFile(), Interpreter.Options().apply {
            numThreads = 2
            useNNAPI   = true   // use Android Neural Networks API if available
        })
    }

    // Pre-computed anchors: [numAnchors × 4] as [cx, cy, w, h] in [0,1]
    private val anchors: Array<FloatArray> by lazy { generateAnchors() }

    // Reusable input buffer: [1, 128, 128, 3] float32
    private val inputBuffer: ByteBuffer = ByteBuffer
        .allocateDirect(1 * INPUT_SIZE * INPUT_SIZE * 3 * 4)
        .order(ByteOrder.nativeOrder())

    override fun callback(frame: Frame, arguments: Map<String, Any>?): Any {
        val image = frame.image

        // Convert YUV_420_888 → RGB Bitmap → scale to 128×128
        val bitmap = yuv420ToBitmap(image)
            ?: return mapOf("detections" to emptyList<Any>())

        val scaled = Bitmap.createScaledBitmap(bitmap, INPUT_SIZE, INPUT_SIZE, true)

        // Fill input buffer: normalise to [-1, 1]
        inputBuffer.rewind()
        for (y in 0 until INPUT_SIZE) {
            for (x in 0 until INPUT_SIZE) {
                val px = scaled.getPixel(x, y)
                inputBuffer.putFloat(((px shr 16 and 0xFF) / 127.5f) - 1f) // R
                inputBuffer.putFloat(((px shr 8  and 0xFF) / 127.5f) - 1f) // G
                inputBuffer.putFloat(((px        and 0xFF) / 127.5f) - 1f) // B
            }
        }

        // Run inference
        val regressors = Array(1) { Array(NUM_ANCHORS) { FloatArray(NUM_COORDS) } }
        val scores     = Array(1) { Array(NUM_ANCHORS) { FloatArray(1) } }
        interpreter.runForMultipleInputsOutputs(
            arrayOf(inputBuffer),
            mapOf(0 to regressors, 1 to scores)
        )

        // Decode detections
        val detections = decode(regressors[0], scores[0], frame.width, frame.height)

        bitmap.recycle()
        scaled.recycle()

        return mapOf("detections" to detections)
    }

    // ─── Decoding ─────────────────────────────────────────────────────────────

    private fun decode(
        regressors: Array<FloatArray>,
        scores: Array<FloatArray>,
        frameW: Int,
        frameH: Int,
    ): List<Map<String, Any>> {
        val raw = mutableListOf<Map<String, Any>>()
        val scaleX = frameW.toFloat() / INPUT_SIZE
        val scaleY = frameH.toFloat() / INPUT_SIZE

        for (i in 0 until NUM_ANCHORS) {
            val score = sigmoid(scores[i][0])
            if (score < SCORE_THRESH) continue

            val r  = regressors[i]
            val ax = anchors[i][0]
            val ay = anchors[i][1]

            val cx = r[0] / INPUT_SIZE + ax
            val cy = r[1] / INPUT_SIZE + ay
            val w  = r[2] / INPUT_SIZE
            val h  = r[3] / INPUT_SIZE

            val kps = (0 until 6).map { k ->
                mapOf(
                    "x" to ((r[4 + k * 2]     / INPUT_SIZE + ax) * scaleX).toDouble(),
                    "y" to ((r[4 + k * 2 + 1] / INPUT_SIZE + ay) * scaleY).toDouble(),
                )
            }

            raw.add(mapOf(
                "x"         to ((cx - w / 2) * scaleX).toDouble(),
                "y"         to ((cy - h / 2) * scaleY).toDouble(),
                "width"     to (w * scaleX).toDouble(),
                "height"    to (h * scaleY).toDouble(),
                "score"     to score.toDouble(),
                "keypoints" to kps,
            ))
        }

        return nonMaxSuppression(raw)
    }

    private fun nonMaxSuppression(dets: List<Map<String, Any>>): List<Map<String, Any>> {
        val sorted = dets.sortedByDescending { it["score"] as Double }
        val kept   = mutableListOf<Map<String, Any>>()
        val suppressed = BooleanArray(sorted.size)

        for (i in sorted.indices) {
            if (suppressed[i]) continue
            kept.add(sorted[i])
            for (j in i + 1 until sorted.size) {
                if (!suppressed[j] && iou(sorted[i], sorted[j]) > IOU_THRESH) {
                    suppressed[j] = true
                }
            }
        }
        return kept
    }

    private fun iou(a: Map<String, Any>, b: Map<String, Any>): Float {
        val ax1 = (a["x"] as Double).toFloat()
        val ay1 = (a["y"] as Double).toFloat()
        val ax2 = ax1 + (a["width"]  as Double).toFloat()
        val ay2 = ay1 + (a["height"] as Double).toFloat()
        val bx1 = (b["x"] as Double).toFloat()
        val by1 = (b["y"] as Double).toFloat()
        val bx2 = bx1 + (b["width"]  as Double).toFloat()
        val by2 = by1 + (b["height"] as Double).toFloat()

        val ix = max(0f, min(ax2, bx2) - max(ax1, bx1))
        val iy = max(0f, min(ay2, by2) - max(ay1, by1))
        val inter = ix * iy
        val union = (ax2-ax1)*(ay2-ay1) + (bx2-bx1)*(by2-by1) - inter
        return inter / (union + 1e-8f)
    }

    // ─── Helpers ─────────────────────────────────────────────────────────────

    private fun sigmoid(x: Float) = (1f / (1f + exp(-x.toDouble()))).toFloat()

    private fun loadModelFile(): MappedByteBuffer {
        val afd: AssetFileDescriptor = proxy.context.assets.openFd(MODEL_FILE)
        val fis = FileInputStream(afd.fileDescriptor)
        return fis.channel.map(FileChannel.MapMode.READ_ONLY, afd.startOffset, afd.declaredLength)
    }

    private fun yuv420ToBitmap(image: android.media.Image): Bitmap? {
        if (image.format != ImageFormat.YUV_420_888) return null
        val yPlane = image.planes[0]
        val uPlane = image.planes[1]
        val vPlane = image.planes[2]

        val yBuffer = yPlane.buffer
        val uBuffer = uPlane.buffer
        val vBuffer = vPlane.buffer

        val ySize = yBuffer.remaining()
        val uSize = uBuffer.remaining()
        val vSize = vBuffer.remaining()

        val nv21 = ByteArray(ySize + uSize + vSize)
        yBuffer.get(nv21, 0, ySize)
        vBuffer.get(nv21, ySize, vSize)
        uBuffer.get(nv21, ySize + vSize, uSize)

        val yuvImage = YuvImage(nv21, ImageFormat.NV21, image.width, image.height, null)
        val out = ByteArrayOutputStream()
        yuvImage.compressToJpeg(Rect(0, 0, image.width, image.height), 80, out)
        val jpegBytes = out.toByteArray()
        return android.graphics.BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.size)
    }

    /** Generate BlazeFace short-range anchors (matches generate_anchors.py output). */
    private fun generateAnchors(): Array<FloatArray> {
        val strides    = intArrayOf(8, 16, 16, 16)
        val anchors    = mutableListOf<FloatArray>()
        val anchorOffsetX = 0.5f
        val anchorOffsetY = 0.5f

        for (layerId in strides.indices) {
            val stride   = strides[layerId]
            val featureW = (INPUT_SIZE + stride - 1) / stride
            val featureH = featureW
            val numAnchorsPerCell = 2

            for (y in 0 until featureH) {
                for (x in 0 until featureW) {
                    repeat(numAnchorsPerCell) {
                        val cx = (x + anchorOffsetX) / featureW
                        val cy = (y + anchorOffsetY) / featureH
                        anchors.add(floatArrayOf(cx, cy, 1f, 1f))
                    }
                }
            }
        }
        return anchors.toTypedArray()
    }
}
